From 095806dd05be556308829d9c42a9a28d158271f0 Mon Sep 17 00:00:00 2001
From: Prarit Bhargava <prarit@redhat.com>
Date: Mon, 14 Jul 2014 13:07:49 -0400
Subject: [powerpc] sysfs: Fix CPU hotplug callback registration

Message-id: <1405343292-21004-13-git-send-email-prarit@redhat.com>
Patchwork-id: 85745
O-Subject: [RHEL7.1 PATCH BZ 1119078 12/35] powerpc, sysfs: Fix CPU hotplug callback registration
Bugzilla: 1119078
RH-Acked-by: Dean Nelson <dnelson@redhat.com>
RH-Acked-by: Igor Mammedov <imammedo@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1119078

commit d1a5511390b0eb242c70ab977abff28644f66a5a
Author: Srivatsa S. Bhat <srivatsa.bhat@linux.vnet.ibm.com>
Date:   Tue Mar 11 02:06:31 2014 +0530

    powerpc, sysfs: Fix CPU hotplug callback registration

    Subsystems that want to register CPU hotplug callbacks, as well as perform
    initialization for the CPUs that are already online, often do it as shown
    below:

     get_online_cpus();

     for_each_online_cpu(cpu)
      init_cpu(cpu);

     register_cpu_notifier(&foobar_cpu_notifier);

     put_online_cpus();

    This is wrong, since it is prone to ABBA deadlocks involving the
    cpu_add_remove_lock and the cpu_hotplug.lock (when running concurrently
    with CPU hotplug operations).

    Instead, the correct and race-free way of performing the callback
    registration is:

     cpu_notifier_register_begin();

     for_each_online_cpu(cpu)
      init_cpu(cpu);

     /* Note the use of the double underscored version of the API */
     __register_cpu_notifier(&foobar_cpu_notifier);

     cpu_notifier_register_done();

    Fix the sysfs code in powerpc by using this latter form of callback
    registration.

    Cc: Benjamin Herrenschmidt <benh@kernel.crashing.org>
    Cc: Paul Mackerras <paulus@samba.org>
    Cc: Olof Johansson <olof@lixom.net>
    Cc: Wang Dongsheng <dongsheng.wang@freescale.com>
    Cc: Ingo Molnar <mingo@kernel.org>
    Acked-by: Madhavan Srinivasan <maddy@linux.vnet.ibm.com>
    Signed-off-by: Srivatsa S. Bhat <srivatsa.bhat@linux.vnet.ibm.com>
    Signed-off-by: Rafael J. Wysocki <rafael.j.wysocki@intel.com>

Cc: Stefan Assmann <sassmann@redhat.com>
Cc: Tony Camuso <tcamuso@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/sysfs.c b/arch/powerpc/kernel/sysfs.c
index b4e6676..6f4984f 100644
--- a/arch/powerpc/kernel/sysfs.c
+++ b/arch/powerpc/kernel/sysfs.c
@@ -657,7 +657,8 @@ static int __init topology_init(void)
  int cpu;
 
  register_nodes();
- register_cpu_notifier(&sysfs_cpu_nb);
+
+ cpu_notifier_register_begin();
 
  for_each_possible_cpu(cpu) {
   struct cpu *c = &per_cpu(cpu_devices, cpu);
@@ -681,6 +682,11 @@ static int __init topology_init(void)
   if (cpu_online(cpu))
    register_cpu_online(cpu);
  }
+
+ __register_cpu_notifier(&sysfs_cpu_nb);
+
+ cpu_notifier_register_done();
+
 #ifdef CONFIG_PPC64
  sysfs_create_dscr_default();
 #endif /* CONFIG_PPC64 */
-- 
1.7.1